### PR TITLE
[HOLD FOR APPROVAL] Add nextActivity check to sqlitenode::update

### PR DESCRIPTION
**Pull request by mcnamamj**
Hi :wave:

This change throttles sqlitenode::update to prevent it from processing all commands in its queues without coming up for air and checking its socket for additional work (that may be higher priority).

This resolves https://github.com/Expensify/Expensify/issues/42345 , which had the symptom where a high volume of low priority commands were being created very quickly and were being processed before normal priority commands.  In reality it was just that the normal priority commands were never getting added to the command queues and so never had a chance.

While I've compiled this code and ran it locally, I'm not in a position where it's easy for me to trigger this and cause it to run.

I've intentionally placed this check *after* the first command runs, so we will always at least process one command.  If there is something in the logic before we get to command processing that consumes the _nextActivity window, we'll at least continue processing some number of commands per second instead of zero.

Something that we should probably discuss is what should be the size of _nextActivity.  It's currently  set to 1 second *from the moment we started reading from the socket*, which seems like during a high load period we may consume most of the time before we even get the chance to start working on commands.

I'd propose that we update https://github.com/Expensify/Bedrock/blob/master/BedrockServer.cpp#L225 to be set to 1 second from STIME_NOW() before this while loop.

```cpp
            // Let the node process any new commands we've opened or existing
            // commands outstanding
+           nextActivity = STimeNow() + STIME_US_PER_S; 
            while (node.update(nextActivity))
                ;   
```

### CODE REVIEW

**Comment by tylerkaraszewski** [comment]
I'm going to test this and make sure it does what it's supposed to.

File: sqlitecluster/SQLiteNode.cpp [hunk 1]
```diff
@@ -1408,6 +1408,14 @@ bool SQLiteNode::update(uint64_t& nextActivity) {
                                 _currentCommand = 0;
                             }
 
+                            // If we've spent enough time processing commands, then it's time to peek/read
+                            // new commands from the socket.  https://github.com/Expensify/Expensify/issues/42345
+                            if (STimeNow() > nextActivity) {
+                                SINFO("Timeout reached while processing commands.  Exceeded by "
+                                    << (STimeNow() - nextActivity) << "ms");
```
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteNode.cpp]: Pretty sure STimeNow() returns µs.

**Comment by tylerkaraszewski** [comment]
I'm testing this and finding some interesting results (also, that the change as-is doesn't work). I'm going to dinner shortly, but will post my complete findings after dinner, which should be before @mcnamamj is back in the morning European time.

**Comment by tylerkaraszewski** [comment]
Second PR added. It's only been superficially tested, but seems to work (i.e., commands keep processing after interrupting because the time limit's been hit).

**Comment by quinthar** [comment]
I still don't see how that can be true.  There is no scenario in which we
queue up 20000 `SendUnreimbursedReportRemindersToManagers` commands,
because each command only queues up to *one* additional command -- and then
exits.  So at any point in time, there is only at most one
`SendUnreimbursedReportRemindersToManagers` in the queue.

Now, I could imagine that it queues it and then immediately executes it --
and if other commands are being added to the queue but not getting
processed, this could cause it to get large.  And if it's true that the
"queue this to run 100ms from now" feature isn't working, that seems
exactly what would happen.  However, if it's in a tight loop adding rows to
the notifications table, and if the problem is it's not checking its
sockets, then clearly it's not creating new higher priority commands.

Rather, if the "queue 100ms in the future" feature isn't working, then what
must be happening is we get into a tight loop creating notifications, and
don't even look at our sockets until all that is done.  And then we realize
"whoa, we've got a ton of work to do" and queue it all up -- and bam, we're
hit by long queue warnings *after SendUnreimbursedReportRemindersToManagers
is all finished*.

Accordingly, it seems like the only change strictly required is to fix the
"queue 100ms in the future" feature -- that alone should fix all this.

Additionally, while I do like the idea of "check the head of the queue to
see if anything higher priority has been added, and if not, continue at the
current position" logic, if we fix the the 100ms delay then this shouldn't
really be necessary.

Finally, I really don't like the idea of breaking the loop and forcing it
to "come up for air" after come arbitrary interval of work.  If there is
work to be done, I think it's more important to immediately do that work,
than to go out looking for more -- even if there is a chance that higher
priority work might be discovered.  It all sounds good in theory, but this
is deep surgery in an area that has worked very well for a long time, and
thus I'm more worried about the unintended consequences of extraneous
changes than the unclear benefits of good sounding changes that aren't
strictly needed.

Basically, I suggest we just fix the "queue 100ms in the future" bug and
avoid making any other changes that aren't strictly necessary.

(Sorry, on a crappy connection so I can't look at the PRs right now, and
thus am making a lot of assumptions.)

**Comment by quinthar** [comment]
(whoops, commented on the wrong issue; moving over to https://github.com/Expensify/Expensify/issues/42345)
